### PR TITLE
Misc updates on master before GAN. Added multithreading + mmod face detector

### PR DESCRIPTION
**Pull request by Clorr**
In order to progress on GAN release, I'm pushing a couple of non GAN related changes.

This release brings multithreading, like what @iSevenDays suggested in #62

The code is mostly clean but could be improved, so feel free to test and give a feedback

### CODE REVIEW

**Comment by Clorr** [comment]
Now with mmod_human_face_detector! Add the -D cnn argument to activate it. But be careful, it is really slow and time consuming, but it detects faces from many more angles

**Comment by gdunstone** [comment]
Is the accuracy higher for this one? And can we specify this for the convert as well? I have had some issues with inaccurate scaling with the current detector for convert but I have processing power up the wazoo.

**Comment by facepainter** [comment]
tl;dr - I get a  "NoneType is not callable" when extracting

My python is not the strongest, need to RTFM myself but to me the global use in #109 lib/multithreading.py seems odd.

Can runner call (processFiles) via the 'method' global? I thought it was not possible to use a global between two separate processes like this... If I do a quick test of this... indeed runner throws "NoneType is not callable".

Again - don't quote me - but pretty sure you need Queue or Pipe, to share picklable stuff between processes... and as this is a bound method I'm not sure that would work.

Maybe I'm too tired or not high enough...totally disregard if this makes no sense - I could well be doing something odd :)

**Comment by Clorr** [comment]
@gdunstone the mmod face detector is in extract & convert, it wouldn't make sense otherwise ;-)

@seattleg yes, the mmod detector is ultra slow on large images, but it remains an alterntive, it is not used by default

@facepainter I'm not really satisfied of how I handled `lib/multithreading.py` but I got it working like that.. The `global method` shouldn't be a big deal because it is written once, and read multiple times, which should be thread-safe. If anyone has a better solution, I'm fine, it also why I did a separate file so that it is simple to modify it

**Comment by gdunstone** [comment]
What's the benefit of using Process over Pool in this instance? It seems like it adds some unnecessary complexity.

**Comment by Clorr** [comment]
Hmm, can you point out which line it is? In `lib/multithreading.py`, I'm using `Pool` because it was almost a one liner, but I had to use an intermediate var because the `map/imap` functions where not ok with a class method

**Comment by gdunstone** [comment]
whoops my bad, I was going off the old version proposed here:

https://github.com/deepfakes/faceswap/pull/62/commits/c4a65ebc147d39105b1acee5345ec756a403b9b7

**Comment by facepainter** [comment]
@Clorr - Hmm yes I still get the error, I get 0 results and the following in the console.

```
Traceback (most recent call last):
  File "U:\faceswap\faceswap.py", line 24, in <module>
    arguments.func(arguments)
  File "U:\faceswap\lib\cli.py", line 53, in process_arguments
    self.process()
  File "U:\faceswap\scripts\extract.py", line 30, in process
    self.faces_detected = sum(pool_process(self.processFiles, list(self.read_directory())))
  File "U:\faceswap\lib\multithreading.py", line 9, in pool_process
    for i in pool.imap_unordered(runner, data):
  File "U:\Python36\Lib\multiprocessing\pool.py", line 735, in next
    raise value
TypeError: 'NoneType' object is not callable
```

> The global method shouldn't be a big deal because it is written once, and read multiple times, which should be thread-safe.

Not pretending to know loads about this or anything, like I said my python is well rusty ( that sounds wrong) ...anyway, basically multi-threading!=multi-processing. Threads share the same memory, processes don't - they have....different process ID, etc. It is confusing because 

```
def runner(item):
    return method(item) 
```

is right there....but "method" here is not "global method" above - this is because when runner is called ```pool.imap_unordered(runner, data)``` it starts a whole new process that has its own memory. At least as I understand it. Or else am I totally missing the point!?! Was pretty sure that is why it was failing for me.... Tell me to shut up if I am missing something silly :) Just really want it to be fassssst :)

**Comment by Clorr** [comment]
@facepainter I see from your stack that you are using python 3.6, while I'm using 3.5, maybe the behavior have change in between. I took reference [here](https://stackoverflow.com/questions/423379/using-global-variables-in-a-function-other-than-the-one-that-created-them) for the `global`.

Also on the thread/process questions, I took what I found in Python, as I'm still beginner in Python, so maybe it is not the right way to go. If you have better samples to show me, feel free.

**Comment by Clorr** [comment]
@facepainter if you find a patch for your platform, let me know. I'm checking on my side, but I don't want to upgrade python for now...

**Comment by Clorr** [comment]
I tried with

```
from concurrent.futures import ThreadPoolExecutor

executor = ThreadPoolExecutor()

def pool_process(method_to_run, data):
    for i in executor.map(method_to_run, data):
        yield i if i is not None else 0
```

And it seems to perform as well as the previous solution. Can you tell me if it is ok for you @facepainter ?

**Comment by facepainter** [comment]
@Clorr

>   I took reference here for the global.

Ah yeah - that wont work across process at all. 
If you imagine, you can't spawn a process - then have that process do a bunch of stuff in memory on another process (i.e. the one that spawned it), you have to use IPC rather than shared memory when doing such things. 

multi threading/processing has a bunch of caveats to it - mainly that anything you pass across processes has to be picklable (serialisable). 

(As an aside this is actually where electron/nodejs excels - it is totally built on the asynchronous IPC model - almost enforces it throughout...more my background too).

Will test now with the thread pool but at a glance that looks good :)

**Comment by Clorr** [comment]
> If it worked for you and you are linux? Then it must be linux/windows os.fork BS. Wouldn't be surprised if global behaves differently here somehow across platform :(
Yes it is annoying

For the multi processing, the sharing of data is a problem, but here, we don't really have to share data, because each task handles one image and saves it, so it simplifies things.

**Comment by facepainter** [comment]
@Clorr  yeah threading is the way to do it, much simpler and avoids the memory isolation / pickling issues :) - that is really good in the profiler too, much faster :)

OOI what are you using to profile/benchmark - i've been using the Python Profiling in Visual Studio 2017  and I'm really impressed.

**Comment by Clorr** [comment]
Ok, let's go for that.

I'm just benchmarking by running an extract with 400 images

**Comment by facepainter** [comment]
lol just RTFM now - https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing.dummy

there is actually a ThreadPool in multiprocessing.  Just not well doc'd at all - not this but something very like...

Using imap_unordered (like you had with the processes) is going to be hella faster than ThreadPoolExecutor.map - plus this was there  is no dependence of futures :) Still, it could be a lot prettier sans globals here....

Really don't want to tread on toes or offend here :) I'm happy to play with this for a couple of hours to get is solid?

```
from multiprocessing.pool import ThreadPool

method = None

def pool_process(method_to_run, data):
    global method
    method = method_to_run
    pool = ThreadPool()
    for i in pool.imap_unordered(runner, data):
        yield(i)
    
def runner(item):
    return method(item)
```

**Comment by gdunstone** [comment]
I profiled these two as ive never seen ThreadPool before. Its hella slow in comparison to a normal Pool
I think its because pythons threading isnt real threading (thanks GIL), and multiprocessing actually has the ability to do real multiprocessing.

```python3

import multiprocessing as mp
from multiprocessing.pool import ThreadPool

method = None

def pool_process(method_to_run, data, j=None):
    global method
    if j is None:
        j = mp.cpu_count() * 2 # I though *2 for HT if you have intel, made no difference
    method = method_to_run
    pool = ThreadPool(threads)
    for i in pool.imap_unordered(runner, data):
        yield(i)

def runner(item):
    return method(item)
```

**Comment by facepainter** [comment]
@gdunstone  - Ah I didn't figure on that overhead - was just thinking imap_unordered vs map :( ...There isn't much info on  ThreadPool at all :(

**Comment by gdunstone** [comment]
Its not that there is an overhead, but that Thread is not **full parallelism**, just concurrent.

https://code.tutsplus.com/articles/introduction-to-parallel-and-concurrent-programming-in-python--cms-28612

**Comment by facepainter** [comment]
@gdunstone - Ah interesting link - thanks :) I really just meant the cost difference of the alternate approaches really - again not claiming to be an expert or anything.

My take away is whilst multiprocessing is not affected by GIL, you cant use non picklable objects...and that there are pretty huge differences between Linux and Windows when it comes to processes, shared memory, and concurrency mechanisms in python (and in general really).

**Comment by Clorr** [comment]
Unless other complaints, I will merge this tomorrow so we can move on

**Comment by facepainter** [comment]
@Clorr  hey totally wasn't a complaint at all - just an observation that using globals across processes won't work on windows :) didn't mean for it to get long winded or to offend :)

**Comment by Clorr** [comment]
No problem, it is just a way to say "feedback" ;-)

**Comment by gdunstone** [comment]
@Clorr Great work on https://github.com/deepfakes/faceswap/pull/109/commits/c83f800c11983147d8289479c2cb06e45ec50f80

**Comment by Apollo122** [comment]
Hey guys,
Have anyone tested this PR on Windows? I can take a look if not.

**Comment by Clorr** [comment]
It would be nice, because even if someone has, maybe he hasn't the same config as you

**Comment by Apollo122** [comment]
Alright then.
Is there any specific config should I use? Or just regular extract, train and convert?

**Comment by Clorr** [comment]
Just regular commands. You can also try different plugins, but that's not mandatory; more specific cases may be raised as issues

**Comment by Apollo122** [comment]
Extract doesnt work. 
`multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "C:\IntelPython3\lib\multiprocessing\pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "K:\faceswap-master\lib\multithreading.py", line 16, in runner
    return method(item)
TypeError: 'NoneType' object is not callable
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "faceswap.py", line 29, in <module>
    arguments.func(arguments)
  File "K:\faceswap-master\lib\cli.py", line 50, in process_arguments
    self.process()
  File "K:\faceswap-master\scripts\extract.py", line 46, in process
    for _ in tqdm(pool_process(self.processFiles, files, processes=processes), total = len(files)):
  File "C:\IntelPython3\lib\site-packages\tqdm\_tqdm.py", line 955, in __iter__
    for obj in iterable:
  File "K:\faceswap-master\lib\multithreading.py", line 12, in pool_process
    for i in pool.imap_unordered(runner, data):
  File "C:\IntelPython3\lib\multiprocessing\pool.py", line 735, in next
    raise value
TypeError: 'NoneType' object is not callable`

**Comment by Clorr** [comment]
Ah yes, this has been discussed above. You can replace the code given by @gdunstone in it's comment https://github.com/deepfakes/faceswap/pull/109#issuecomment-362954373 The code goes in `lib/multithreading.py`. If it doesn't work let us know, or try another of the versions above

**Comment by Apollo122** [comment]
I got the same error for both codes above.

Loading Extract from Extract_Align plugin...
Traceback (most recent call last):
  File "faceswap.py", line 29, in <module>
    arguments.func(arguments)
  File "K:\faceswap-master\lib\cli.py", line 50, in process_arguments
    self.process()
  File "K:\faceswap-master\scripts\extract.py", line 46, in process
    for _ in tqdm(pool_process(self.processFiles, files, processes=processes), total = len(files)):
TypeError: pool_process() got an unexpected keyword argument 'processes'

**Comment by Clorr** [comment]
Oh, it is because of the signature of the copy/paste'd code does not have the argument `processes`. It is named `j` in the copy/paste'd code I showed you, it should be renamed. Test if it is still ok for you, and maybe push a correction, if not I will finalize on my side tomorrow

**Comment by Apollo122** [comment]
So this one didnt work
python faceswap.py extract -i photos/A -o data/A --processes 2

but this worked: python faceswap.py extract -i photos/A -o data/A --processes 1

but its not utilizing multi-threading anymore

**Comment by Clorr** [comment]
Yes, because with 1 process, it does not call the 'bad' part with the `processes` argument. But if other parts are ok, let me know

**Comment by Apollo122** [comment]
Ok i changed both "j" and "threads" with "processes" and its working. 
i tried with --processes 8 but it still takes same time. I didn't see any improvement with remaining time.

**Comment by Clorr** [comment]
hmm strange. Have you tried opening the resource monitor to see if all cores are used? also, how big is your data set ? If it is too small, the startup overhead may compensate the parallelization

**Comment by Apollo122** [comment]
yeah i see the splikes. 
on task manager, it shows under CPU python.exe uses 13.
Im extracting 640 images

**Comment by Clorr** [comment]
13% ? If you have 4 cores + hyperthreading, it is likely that it uses only one core (100% / 8 = 12.5%)

640 images is largely enough for the test.

I'll have more in-depth look about it tomorrow...

**Comment by Apollo122** [comment]
i have 4 cores yes. 
btw i tried train got this error and is stuck.
Unable to create file (unable to open file: name = 'models/encoder.h5', errno = 2, error message = 'No such file or directory', flags = 13, o_flags = 302)

created the folder manually and it worked.

**Comment by Clorr** [comment]
Ah thanks for this bug. maybe I should add a check for the output folder

**Comment by Apollo122** [comment]
So i tested the convert as well seems fine.
Although I have a suggestion about it:

Right now convert takes a full frame. First it detects the face then crops it, runs model and finally merges it back.
But we already done half the job during extracting, we detected the face and cropped it to train on it. So it doesn't make sense to do it all over again(especially if we want to use cnn). Also algorithm may miss some faces or output some garbage images. With this approach we have absolutely no control what images it uses. We should use the extracted images to process and merge those on the frame. For example 1adrianb's face-detector outputs alignment.json after extraction. With using something like that we would know where the face is on the frame and where to merge it without processing again.

**Comment by facepainter** [comment]
@Clorr have been playing with this myself - the face_detection lib actually allows you to batch process upto 128 frames at a time using a CNN model. You just have to make sure that you compile dlib with CUDA extensions and then you get to use the GPU too :) totally removes the ffmpeg splitting to frames step :)  (oh - all the frames do have to be the exact same dimensions!).

I have just been testing it - the idea is you give it a video, a small sample of images of a person you want (say a directory with 5-10 good shots). an output folder and a batch size i.e.

`faceswap.py extract -i videoOfX.mp4 -r ./ImagesOfX -o ./grabedFramesX -b 32`

It is actually working pretty well - it is faster - and as it negates first extracting and loading the frames as a step outside the application in ffmpeg or whatever. 

I even put together a version with youtube-dl included so you can do...

`faceswap.py extract -i 2z3JUS1aXdY  -r ./ImagesOfX -o ./grabedFramesX -b 32`

Where **2z3JUS1aXdY**  is the id of a youtube video (it just gets the best version, then does as above).

Really just all just based off this example: https://github.com/ageitgey/face_recognition/blob/master/examples/find_faces_in_batches.py

Anyhow thought you might be interested - apologies if this is something that has already been looked at :)

**Comment by Clorr** [comment]
@Apollo122 supporting an `alignment.json` file is totally feasible because it is just a matter of saving the landmarks that are generated on extract. Just this is not related to this PR so maybe discuss it in a specific issue.

@facepainter if you have some code to propose, we are totally open to it ;-) Feel free to make a PR. Just be careful to make it after this one has been merged, so we don't get too much conflicts

**Comment by Apollo122** [comment]
I classified that as misc update also :)
But sure can be opened as different issue

**Comment by Clorr** [comment]
I'm merging this now because we should move on, don't hesitate to open issues if you find bugs...

**Comment by modelsex** [comment]
cnn mode work perfect on Extract, but I get follow problem when use cnn to convert:

Command example: python faceswap.py convert -i D:/project4/data_A/ -o D:/face/final/ -m D:/face/project4/models/ -D cnn 

My Env. CUDA-8.0 | cudnn-8.0 | dlib-19.9

**Log output**
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.7715
pciBusID: 0000:01:00.0
totalMemory: 8.00GiB freeMemory: 6.62GiB
2018-02-08 19:24:41.501735: I C:\tf_jenkins\home\workspace\rel-win\M\windows-gpu\PY\36\tensorflow\core\common_runtime\gpu\gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 1080, pci bus id: 0000:01:00.0, compute capability: 6.1)
loaded model weights
Loading Convert from Convert_Masked plugin...
  0%|                                                                                          | 0/800 [00:00<?, ?it/s]Failed to convert image: D:\project4\data_A\kar0001.png. **Reason: Error while calling cudaMalloc(&data, n) in file D:\dlib-master\dlib\dnn\cuda_data_ptr.cpp:28. code: 2, reason: out of memory**
  0%|▏                                                                                 | 2/800 [00:00<04:24,  3.02it/s]Failed to convert image: D:\project4\data_A\kar0002.png. **Reason: Error while calling cudaGetLastError() in file D:\dlib-master\dlib\dnn\gpu_data.cpp:91. code: 2, reason: out of memory**
Failed to convert image: D:\project4\data_A\kar0003.png. Reason: Error while calling cudaMalloc(&data, n) in file D:\dlib-master\dlib\dnn\cuda_data_ptr.cpp:28. code: 2, reason: out of memory
  0%|▍                                                                                 | 4/800 [00:00<02:34,  5.16it/s]Failed to convert image: D:\project4\data_A\kar0004.png. Reason: Error while calling cudaGetLastError() in file D:\dlib-master\dlib\dnn\gpu_data.cpp:91. code: 2, reason: out of memory
......

**Comment by Clorr** [comment]
It is because convert loads cnn **+** model, and for some reason they take both much memory. I have added a patch in #140 that tells tensorflow to allocate memory on-demand and not all at once. I don't gaurantee it works, but if you can try and give feedback it would be nice....
